asset-transfer-events/chaincode-javascript/lib/assetTransferEvents.js

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright IBM Corp. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict';
+
+const { Contract } = require('fabric-contract-api');
+
+async function savePrivateData(ctx, assetKey) {
+	const clientOrg = ctx.clientIdentity.getMSPID();
+	const peerOrg = ctx.stub.getMspID();
+	const collection = '_implicit_org_' + peerOrg;
+
+	if (clientOrg === peerOrg) {
+		const transientMap = ctx.stub.getTransient();
+		if (transientMap) {
+			const properties = transientMap.get('asset_properties');
+			if (properties) {
+				await ctx.stub.putPrivateData(collection, assetKey, properties);
+			}
+		}
+	}
+}
+
+async function removePrivateData(ctx, assetKey) {
+	const clientOrg = ctx.clientIdentity.getMSPID();
+	const peerOrg = ctx.stub.getMspID();
+	const collection = '_implicit_org_' + peerOrg;
+
+	if (clientOrg === peerOrg) {
+		const propertiesBuffer = await ctx.stub.getPrivateData(collection, assetKey);
+		if (propertiesBuffer && propertiesBuffer.length > 0) {
+			await ctx.stub.deletePrivateData(collection, assetKey);
+		}
+	}
+}
+
+async function addPrivateData(ctx, assetKey, asset) {
+	const clientOrg = ctx.clientIdentity.getMSPID();
+	const peerOrg = ctx.stub.getMspID();
+	const collection = '_implicit_org_' + peerOrg;
+
+	if (clientOrg === peerOrg) {
+		const propertiesBuffer = await ctx.stub.getPrivateData(collection, assetKey);
+		if (propertiesBuffer && propertiesBuffer.length > 0) {
+			const properties = JSON.parse(propertiesBuffer.toString());
+			asset.asset_properties = properties;
+		}
+	}
+}
+
+async function readState(ctx, id) {
+	const assetBuffer = await ctx.stub.getState(id); // get the asset from chaincode state
+	if (!assetBuffer || assetBuffer.length === 0) {
+		throw new Error(`The asset ${id} does not exist`);
+	}
+	const assetString = assetBuffer.toString();
+	const asset = JSON.parse(assetString);
+
+	return asset;
+}
+
+class AssetTransferEvents extends Contract {
+
+	// CreateAsset issues a new asset to the world state with given details.
+	async CreateAsset(ctx, id, color, size, owner, appraisedValue) {
+		const asset = {
+			ID: id,
+			Color: color,
+			Size: size,
+			Owner: owner,
+			AppraisedValue: appraisedValue,
+		};
+		await savePrivateData(ctx, id);
+		const assetBuffer = Buffer.from(JSON.stringify(asset));
+
+		ctx.stub.setEvent('CreateAsset', assetBuffer);
```

Comment:
setEvent should probably be sent out after putState is successful

same for TransferEvent

Reply:
doesn't actually matter as the event is not sent, the event data is just included with the transaction data. The peer and then the application does not see this until the block with this transaction gets committed.

Reply:
I will change as it does look better

Reply:
Not able to change as I want to return the status of the putState